File: Tools/DocLuaParserBasic_Lua/ParseFunctions.lua
local Utils = require("Tools.Utils")

---@class Tools.DocLuaParserBasic.ParseFunctions
local ParseFunctions = {}

---@param line string
---@return boolean
function ParseFunctions.IsComment(line)
    return line:find("^%-%-") ~= nil
end

---@param line string
---@return string comment
function ParseFunctions.ParseComment(line)
    return line:match("^%-%-%-? (.*)")
end

---@param line string
---@return boolean
function ParseFunctions.IsFunction(line)
    if line:find("^local ", nil, true) then
        line = line:sub(7, line:len())
    end
    return line:find("^function .+%(.*%)") ~= nil
end

---@param line string
---@return string name, string[] args
function ParseFunctions.ParseFunction(line)
    local name, argsStr = line:match("function (.+)%((.*)%)")
    return name, Utils.SplitStr(argsStr, ", ")
end

---@param line string
---@return boolean
function ParseFunctions.IsParameter(line)
    return line:find("^%-%-%-@param ") ~= nil
end

---@param line string
---@return string name, string type, string description
function ParseFunctions.ParseParameter(line)
    local name, type, description = line:match("^%-%-%-@param (.+) (.+) ?(.*)$")
    return name, type, description
end

---@param line string
---@return boolean
function ParseFunctions.IsReturn(line)
    return line:find("^%-%-%-@return ") ~= nil
end

---@return string name
---@return string type
---@return string description
function ParseFunctions.ParseReturn(line)
    local type, name, description = line:match("^%-%-%-@return (.+) (.+) ?(.*)$")
    return name, type, description
end

return ParseFunctions

File: Tools/DocUpdater/src/ParseFunctions.cs
using System.Diagnostics.CodeAnalysis;
using System.Numerics;
using System.Text.RegularExpressions;

namespace DocUpdater
{
    internal static partial class ParseFunctions
    {
        [GeneratedRegex("^---? ?(.*)")]
        private static partial Regex GetCommentRegex();

        internal static bool IsComment(string line, [MaybeNullWhen(false)] out string comment)
        {
            comment = string.Empty;
            var regex = GetCommentRegex();

            var match = regex.Match(line);
            if (!match.Success)
                return false;

            comment = match.Groups[1].Value;
            return true;
        }


        [GeneratedRegex("^@param (\\S+) (.*?) ?(?:--? ?(.*))?$")]
        private static partial Regex GetParameterRegex();

        internal static bool IsParameter(string line, [MaybeNullWhen(false)] out FuncParameter data)
        {
            data = default;
            var regex = GetParameterRegex();

            var match = regex.Match(line);
            if (!match.Success)
                return false;

            var name = match.Groups[1].Value;
            var type = match.Groups[2].Value.Replace("|", "/");
            var description = match.Groups.Count > 3 ? match.Groups[3].Value : string.Empty;

            data = new() { Name = name, Type = type, Description = description };
            return true;
        }


        [GeneratedRegex("^@return (.+?)(?: ([^|,<> ]+?))? ?(?:--? ?(.*))?$")]
        private static partial Regex GetReturnRegex();

        internal static bool IsReturn(string line, [MaybeNullWhen(false)] out FuncParameter data)
        {
            data = default;
            var regex = GetReturnRegex();

            var match = regex.Match(line);
            if (!match.Success)
                return false;

            var matchGroupsCount = match.Groups.Count;

            var type = match.Groups[1].Value.Replace("|", "/");
            var name = matchGroupsCount > 2 ? match.Groups[2].Value : type;
            var description = matchGroupsCount > 3 ? match.Groups[3].Value : string.Empty;

            data = new() { Name = name, Type = type, Description = description };
            return true;
        }

        
        [GeneratedRegex("(?:local )?function ([^(]+)\\((.*)\\)")]
        private static partial Regex GetFunctionRegex();

        internal static bool IsFunction(string line, [MaybeNullWhen(false)] out (string Name, string[] argNames) data)
        {
            data = default;
            var regex = GetFunctionRegex();

            var match = regex.Match(line);
            if (!match.Success)
                return false;

            var name = match.Groups[1].Value;
            var argNamesStr = match.Groups.Count > 2 ? match.Groups[2].Value : string.Empty;

            var argNames = argNamesStr.Split(", ");
            data = (name, argNames);
            return true;
        }


        [GeneratedRegex("^@class (?<! : )(.+?)(?: : (.*))?$")]
        private static partial Regex GetClassRegex();

        internal static bool IsClass(string line, out string name)
        {
            name = string.Empty;
            var regex = GetClassRegex();

            var match = regex.Match(line);
            if (!match.Success)
                return false;

            name = match.Groups[1].Value;
            return true;
        }
    }
}
